packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -119,23 +187,33 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     )
     internal
     override
-    returns (bytes memory context, bool revertOnRecipientRevert) {
-        (signature);
-        require(approvalData.length == 0, "approvalData: invalid length");
-        if (relayRequest.relayData.paymasterData.length != 0) {
-            require(relayRequest.relayData.paymasterData.length >= 4, "paymastaData: missing method sig");
+    returns (bytes memory, bool) {
+        (signature, approvalData);
+        uint256 paymasterDataLength = relayRequest.relayData.paymasterData.length;
+        // paymasterData must contain the token address, and optionally a a valid "permit" call on the token.
+        require(paymasterDataLength >= 20, "must contain token address");
+        IERC20Metadata token = _getTokenFromPaymasterData(relayRequest.relayData.paymasterData);
+        require(isTokenSupported(token),"unsupported token");
+        TokenSwapData memory tokenSwapData = tokensSwapData[token];
+        if (paymasterDataLength != 20) {
+            require(paymasterDataLength >= 24, "must contain \"permit\" and token");
             require(
-                permitMethodSignature == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
-                "paymasterData: wrong method sig");
+                tokenSwapData.permitMethodSignature == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
+                "wrong \"permit\" method sig");
```

Comment:
I know I wrote the original code like this, but now I guess it would make much more sense to encode on-chain then to decode-and-compare on-chain now, what do you think?

Reply:
You know what? It is fine.